src/QsCompiler/QirGeneration/RewriteSteps.cs

```diff
@@ -4,14 +4,98 @@
+            this.AssemblyConstants = new Dictionary<string, string?>();
+        }
+
+        /// <inheritdoc/>
+        public string Name => "QIR Generation";
+
+        /// <inheritdoc/>
+        public int Priority => EmissionPriority;
+
+        /// <inheritdoc/>
+        public IDictionary<string, string?> AssemblyConstants { get; }
+
+        /// <inheritdoc/>
+        public IEnumerable<IRewriteStep.Diagnostic> GeneratedDiagnostics => this.diagnostics;
+
+        /// <inheritdoc/>
+        public bool ImplementsPreconditionVerification => true;
+
+        /// <inheritdoc/>
+        public bool ImplementsTransformation => true;
+
+        /// <inheritdoc/>
+        public bool ImplementsPostconditionVerification => false;
+
+        /// <inheritdoc/>
+        public bool PreconditionVerification(QsCompilation compilation)
+        {
+            try
+            {
+                ValidateMonomorphization.Apply(compilation);
+                return true;
+            }
+            catch
+            {
+                this.diagnostics.Add(new IRewriteStep.Diagnostic
+                {
+                    Severity = DiagnosticSeverity.Error,
+                    Stage = IRewriteStep.Stage.PreconditionVerification,
+                    Message = DiagnosticItem.Message(ErrorCode.SyntaxTreeNotMonomorphized, Array.Empty<string>())
+                });
+                return false;
+            }
+        }
+
+        /// <inheritdoc/>
+        public bool Transformation(QsCompilation compilation, out QsCompilation transformed)
+        {
+            transformed = compilation;
+            var generator = new Generator(transformed);
+            generator.Apply();
+
+            // write generated QIR to disk
+            var assemblyName = this.AssemblyConstants.TryGetValue(ReservedKeywords.AssemblyConstants.AssemblyName, out var asmName) ? asmName : null;
+            var targetFile = Path.GetFullPath(string.IsNullOrWhiteSpace(assemblyName) ? "main.txt" : $"{Path.GetFileName(assemblyName)}.txt");
+
+            PerformanceTracking.TaskStart(PerformanceTracking.Task.BitcodeGeneration);
+            var bcOutputFolder = this.AssemblyConstants.TryGetValue(ReservedKeywords.AssemblyConstants.OutputPath, out var path) && !string.IsNullOrWhiteSpace(path) ? path : "qir";
+            var bcFile = CompilationLoader.GeneratedFile(targetFile, Path.GetFullPath(bcOutputFolder), ".bc", "");
+            generator.Emit(bcFile, emitBitcode: true);
+            PerformanceTracking.TaskEnd(PerformanceTracking.Task.BitcodeGeneration);
+
+            // create the human readable version as well
+            var sourceOutputFolder = this.AssemblyConstants.TryGetValue(ReservedKeywords.AssemblyConstants.QirOutputPath, out path) && !string.IsNullOrWhiteSpace(path) ? path : "qir";
+            var llvmSourceFile = CompilationLoader.GeneratedFile(targetFile, Path.GetFullPath(sourceOutputFolder), ".ll", "");
```

Comment:
With the other changes, does this mean that the .ll file will no longer be generated in the "qir" folder with a name matching the project, but instead be named "__qsharp_data_qir_v1__.ll" and end up in the regular tree? If so, that will break some logic in scripts in the runtime repo that expect to find the .ll file in a specific place. Now that we are producing bitcode we can probably update those scripts to expect bitcode instead, but we'd need to coordinate a cross-repo check in for this since it would be a breaking change.

Reply:
No, it is still in the same place. The behavior of the Sdk (i.e. the use of the feature within a Q# project) did not change; the only things that is added is that the bc is also attached now.